statistics/handle/handle.go

```diff
@@ -472,21 +478,26 @@ func (h *Handle) tableStatsFromStorage(tableInfo *model.TableInfo, physicalID in
 	}
 	table.Pseudo = false
 	selSQL := fmt.Sprintf("select table_id, is_index, hist_id, distinct_count, version, null_count, tot_col_size, stats_ver, flag, correlation from mysql.stats_histograms where table_id = %d", physicalID)
-	rows, _, err := h.restrictedExec.ExecRestrictedSQL(nil, selSQL)
+	var rows []chunk.Row
+	if historyStatsExec != nil {
+		rows, _, err = historyStatsExec.ExecRestrictedSQLWithSnapshot(nil, selSQL)
+	} else {
+		rows, _, err = h.restrictedExec.ExecRestrictedSQL(nil, selSQL)
+	}
 	if err != nil {
-		return nil, errors.Trace(err)
+		return nil, err
 	}
 	// Check deleted table.
 	if len(rows) == 0 {
 		return nil, nil
 	}
 	for _, row := range rows {
 		if row.GetInt64(1) > 0 {
-			if err := h.indexStatsFromStorage(row, table, tableInfo); err != nil {
+			if err := h.indexStatsFromStorage(row, table, tableInfo, historyStatsExec); err != nil {
```

Comment:
The stats of index or column will only be updated when `idx.LastUpdateVersion < histVer`, so it is better to use an empty `table`, not the one in the stats cache.